pkg/compact/downsample/writer.go

```diff
@@ -0,0 +1,342 @@
+package downsample
+
+import (
+	"encoding/json"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/block"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/tsdb"
+	"github.com/prometheus/tsdb/chunks"
+	"github.com/prometheus/tsdb/fileutil"
+	"github.com/prometheus/tsdb/index"
+	"github.com/prometheus/tsdb/labels"
+)
+
+type symbols map[string]struct{}
+type labelValues map[string]struct{}
+
+func (lv labelValues) add(value string) {
+	lv[value] = struct{}{}
+}
+func (lv labelValues) get(set *[]string) {
+	for value := range lv {
+		*set = append(*set, value)
+	}
+}
+
+type labelsValues map[string]labelValues
+
+func (lv labelsValues) add(labelSet labels.Labels) {
+	for _, label := range labelSet {
+		values, ok := lv[label.Name]
+		if !ok {
+			// Add new label.
+			values = labelValues{}
+			lv[label.Name] = values
+		}
+		values.add(label.Value)
+	}
+}
+
+// InstantWriter writes downsampled block to a new data block. Chunks will be written immediately in order to avoid
+// memory consumption.
+type instantWriter struct {
+	dir        string
+	tmpDir     string
+	logger     log.Logger
+	uid        ulid.ULID
+	resolution int64
+
+	symbols  symbols
+	postings []uint64
+	series   []*series
+
+	chunkWriter  tsdb.ChunkWriter
+	meta         block.Meta
+	totalChunks  uint64
+	totalSamples uint64
+}
+
+func NewWriter(dir string, l log.Logger, originMeta block.Meta, resolution int64) (*instantWriter, error) {
+	var err error
+	var chunkWriter tsdb.ChunkWriter
+
+	// Generate new block id
+	entropy := rand.New(rand.NewSource(time.Now().UnixNano()))
+	uid := ulid.MustNew(ulid.Now(), entropy)
+
+	// Populate chunk, meta and index files into temporary directory with
+	// data of all blocks.
+	dir = filepath.Join(dir, uid.String())
+	tmpDir, err := createTmpDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	chunkDir := func(dir string) string {
```

Comment:
do you need that function? why not just inlining  `filepath.Join(dir, block.ChunksDirname)`

Reply:
no, redundancy was forgotten